[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137723
* #137661
* #137426
* #137722
* __->__ #137377

ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

swolchok committed:
[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2

ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot label "ciflow/linux-aarch64"

swolchok committed:
Update on "[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2"


ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2"


ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

[ghstack-poisoned]

albanD commented:
What's the difference with the previous attempt?

swolchok commented on aten/src/ATen/native/BlasKernel.cpp:
```diff
@@ -455,67 +456,183 @@ static C10_ALWAYS_INLINE float32x4_t f32_fma_bf16(float32x4_t a, uint16x4_t b, u
+  return reduce(sum);
 }
 
+// NOTE: The first attempt at landing BFDOT support with
```
@albanD here is the difference

albanD replied:
Ho thanks for the pointer!
Should we undef DOT_WITH_FP32_ARITH_TAIL_AFTER_MAIN_LOOP_BODY ?

swolchok replied:
this is a .cpp file, so the macro isn't going to leak anywhere and it's not particularly necessary, but sure I can do that.

swolchok replied:
done

swolchok committed:
Update on "[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2"


ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

[ghstack-poisoned]

swolchok commented:
will fix https://hud.pytorch.org/pr/pytorch/pytorch/137377#31249893287

swolchok committed:
more thorough GCC fix on "[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2"


ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

[ghstack-poisoned]

swolchok committed:
undef macro on "[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2"


ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

[ghstack-poisoned]

swolchok commented:
CI is 100% green; please review

swolchok committed:
Update on "[PyTorch] Port ExecuTorch bfdot improvement back to ATen BlasKernel, Try #2"


ExecuTorch's fork of BlasKernel.cpp grew bfdot support, complete with demonstration that it helps. Port it back to PyTorch. First attempt was https://github.com/pytorch/pytorch/pull/136331 .

Differential Revision: [D63923166](https://our.internmc.facebook.com/intern/diff/D63923166/)

[ghstack-poisoned]

malfet commented:
If CI is green, than sure, LGTM

swolchok commented:
@pytorchbot merge
